File: BubbleSort.java
// 始终生效
// 模型决策

/**
 * 冒泡排序实现类
 * 提供多种冒泡排序的实现方法
 */
public class BubbleSort {
    
    /**
     * 基础冒泡排序算法
     * 时间复杂度: O(n²)
     * 空间复杂度: O(1)
     * 
     * @param arr 待排序的整型数组
     */
    public static void bubbleSort(int[] arr) {
        if (arr == null || arr.length <= 1) {
            return;
        }
        
        int n = arr.length;
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - 1 - i; j++) {
                if (arr[j] > arr[j + 1]) {
                    // 交换元素
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }
    }
    
    /**
     * 优化的冒泡排序算法（添加标志位）
     * 如果某次遍历没有发生交换，说明数组已经有序，可以提前结束
     * 最优时间复杂度: O(n)
     * 最坏时间复杂度: O(n²)
     * 
     * @param arr 待排序的整型数组
     */
    public static void bubbleSortOptimized(int[] arr) {
        if (arr == null || arr.length <= 1) {
            return;
        }
        
        int n = arr.length;
        boolean swapped;
        
        for (int i = 0; i < n - 1; i++) {
            swapped = false;
            for (int j = 0; j < n - 1 - i; j++) {
                if (arr[j] > arr[j + 1]) {
                    // 交换元素
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                    swapped = true;
                }
            }
            // 如果本轮没有发生交换，说明已经有序
            if (!swapped) {
                break;
            }
        }
    }
    
    /**
     * 降序冒泡排序
     * 
     * @param arr 待排序的整型数组
     */
    public static void bubbleSortDescending(int[] arr) {
        if (arr == null || arr.length <= 1) {
            return;
        }
        
        int n = arr.length;
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - 1 - i; j++) {
                if (arr[j] < arr[j + 1]) {
                    // 交换元素
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }
    }
    
    /**
     * 打印数组元素
     * 
     * @param arr 要打印的数组
     */
    public static void printArray(int[] arr) {
        if (arr == null) {
            System.out.println("数组为空");
            return;
        }
        
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
            if (i < arr.length - 1) {
                System.out.print(", ");
            }
        }
        System.out.println();
    }
    
    /**
     * 主函数，用于测试冒泡排序
     */
    public static void main(String[] args) {
        // 测试基础冒泡排序
        System.out.println("=== 测试基础冒泡排序 ===");
        int[] arr1 = {64, 34, 25, 12, 22, 11, 90};
        System.out.print("排序前: ");
        printArray(arr1);
        bubbleSort(arr1);
        System.out.print("排序后: ");
        printArray(arr1);
        
        // 测试优化的冒泡排序
        System.out.println("\n=== 测试优化的冒泡排序 ===");
        int[] arr2 = {5, 1, 4, 2, 8};
        System.out.print("排序前: ");
        printArray(arr2);
        bubbleSortOptimized(arr2);
        System.out.print("排序后: ");
        printArray(arr2);
        
        // 测试降序排序
        System.out.println("\n=== 测试降序冒泡排序 ===");
        int[] arr3 = {3, 7, 1, 9, 2};
        System.out.print("排序前: ");
        printArray(arr3);
        bubbleSortDescending(arr3);
        System.out.print("排序后: ");
        printArray(arr3);
        
        // 测试已排序数组（验证优化效果）
        System.out.println("\n=== 测试已排序数组 ===");
        int[] arr4 = {1, 2, 3, 4, 5};
        System.out.print("排序前: ");
        printArray(arr4);
        bubbleSortOptimized(arr4);
        System.out.print("排序后: ");
        printArray(arr4);
    }
}

File: bubble_sort.py
"""
始终生效
冒泡排序实现模块
提供冒泡排序算法的实现和相关功能
"""


def bubble_sort(arr):
    """
    冒泡排序算法实现
    
    参数:
        arr: 待排序的列表
    
    返回:
        排序后的列表
    
    时间复杂度: O(n²)
    空间复杂度: O(1)
    """
    if not arr:
        return arr
    
    n = len(arr)
    # 创建副本以避免修改原列表
    result = arr.copy()
    
    # 外层循环控制排序轮数
    for i in range(n - 1):
        # 标记本轮是否发生交换
        swapped = False
        
        # 内层循环进行相邻元素比较和交换
        for j in range(n - 1 - i):
            if result[j] > result[j + 1]:
                # 交换相邻元素
                result[j], result[j + 1] = result[j + 1], result[j]
                swapped = True
        
        # 如果本轮没有发生交换,说明已经有序,可以提前结束
        if not swapped:
            break
    
    return result


def bubble_sort_descending(arr):
    """
    冒泡排序 - 降序实现
    
    参数:
        arr: 待排序的列表
    
    返回:
        降序排序后的列表
    """
    if not arr:
        return arr
    
    n = len(arr)
    result = arr.copy()
    
    for i in range(n - 1):
        swapped = False
        
        for j in range(n - 1 - i):
            if result[j] < result[j + 1]:  # 改为降序比较
                result[j], result[j + 1] = result[j + 1], result[j]
                swapped = True
        
        if not swapped:
            break
    
    return result


def bubble_sort_with_steps(arr):
    """
    冒泡排序 - 显示每一步的排序过程
    
    参数:
        arr: 待排序的列表
    
    返回:
        元组 (排序后的列表, 排序步骤列表)
    """
    if not arr:
        return arr, []
    
    n = len(arr)
    result = arr.copy()
    steps = [result.copy()]
    
    for i in range(n - 1):
        swapped = False
        
        for j in range(n - 1 - i):
            if result[j] > result[j + 1]:
                result[j], result[j + 1] = result[j + 1], result[j]
                swapped = True
                steps.append(result.copy())
        
        if not swapped:
            break
    
    return result, steps


def demo():
    """演示冒泡排序的使用"""
    print("=" * 50)
    print("冒泡排序演示")
    print("=" * 50)
    
    # 示例1: 基本升序排序
    arr1 = [64, 34, 25, 12, 22, 11, 90]
    print(f"\n原始数组: {arr1}")
    sorted_arr1 = bubble_sort(arr1)
    print(f"升序排序后: {sorted_arr1}")
    
    # 示例2: 降序排序
    arr2 = [5, 2, 8, 1, 9, 3]
    print(f"\n原始数组: {arr2}")
    sorted_arr2 = bubble_sort_descending(arr2)
    print(f"降序排序后: {sorted_arr2}")
    
    # 示例3: 显示排序步骤
    arr3 = [5, 3, 8, 4, 2]
    print(f"\n原始数组: {arr3}")
    sorted_arr3, steps = bubble_sort_with_steps(arr3)
    print(f"排序步骤:")
    for idx, step in enumerate(steps):
        print(f"  步骤 {idx}: {step}")
    print(f"最终结果: {sorted_arr3}")
    
    # 示例4: 已排序数组
    arr4 = [1, 2, 3, 4, 5]
    print(f"\n已排序数组: {arr4}")
    sorted_arr4 = bubble_sort(arr4)
    print(f"排序后: {sorted_arr4}")
    
    # 示例5: 空数组和单元素数组
    print(f"\n空数组: {bubble_sort([])}")
    print(f"单元素数组: {bubble_sort([42])}")


if __name__ == "__main__":
    demo()
